sql: Add support for OWNED BY option in CREATE/ALTER SEQUENCE Commands

arulajmani opened the pull request:
This PR adds sequence ownership semantics to CRDB.

- By default, a sequence is not owned by any column.

- The OWNED BY option allows the user to establish an ownership
relationiship of a sequence  with a particular table's column.

- Ownership relationships can be removed by using "OWNED BY NONE"

Fixes #26382

Release note (sql change): CREATE/ALTER SEQUENCE now support the OWNED BY
syntax.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/40992)
<!-- Reviewable:end -->

solongordon reviewed (changes requested):
Nice start! I still need to review the code around dropping sequences but here are some thoughts to get you started.

> Reviewed 5 of 5 files at r1, 1 of 1 files at r2, 4 of 24 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-:-LpYqzPG4i5KQJckWSgm:b3s3689)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @solongordon)

___
*[pkg/ccl/importccl/read_import_mysql.go, line 269 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-LpYeirt2iVKKdqwaT2Z:-LpYeirt2iVKKdqwaT2_:b-x5rsed) ([raw file](https://github.com/cockroachdb/cockroach/blob/09c2833585e1c68d64dfb35c7deee384e02b01ca/pkg/ccl/importccl/read_import_mysql.go#L269)):*
> ```Go
> func readMysqlCreateTable(
> 	input io.Reader,
> 	ctx context.Context,
> ```

By convention, `ctx` is usually the first argument.
___
*[pkg/ccl/importccl/read_import_mysql.go, line 344 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-LpYfoyo3wKVBTglzSoo:-LpYfoyo3wKVBTglzSop:bgnbm0j) ([raw file](https://github.com/cockroachdb/cockroach/blob/09c2833585e1c68d64dfb35c7deee384e02b01ca/pkg/ccl/importccl/read_import_mysql.go#L344)):*
> ```Go
> 	ctx context.Context,
> 	evalCtx *tree.EvalContext,
> 	p *sql.PlanHookState,
> ```

I don't think this needs to be a pointer. `PlanHookState` is an interface, and the underlying implementation might be a pointer, but using a pointer to an interface is almost always undesirable in Go.
___
*[pkg/ccl/importccl/read_import_pgdump.go, line 212 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-LpYgX0y0MKoMT1EZY5E:-LpYgX0y0MKoMT1EZY5F:bfasxok) ([raw file](https://github.com/cockroachdb/cockroach/blob/09c2833585e1c68d64dfb35c7deee384e02b01ca/pkg/ccl/importccl/read_import_pgdump.go#L212)):*
> ```Go
> func readPostgresCreateTable(
> 	input io.Reader,
> 	ctx context.Context,
> ```

Ditto about making `ctx` the first argument.
___
*[pkg/sql/sequence.go, line 191 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-LpYnPh586gXgtgOesNz:-LpYnPh586gXgtgOesO-:bbhxg0j) ([raw file](https://github.com/cockroachdb/cockroach/blob/09c2833585e1c68d64dfb35c7deee384e02b01ca/pkg/sql/sequence.go#L191)):*
> ```Go
> // starting with defaults and overriding them with user-provided options.
> func assignSequenceOptions(
> 	opts *sqlbase.TableDescriptor_SequenceOpts, optsNode tree.SequenceOptions, setDefaults bool, params *runParams, sequenceID sqlbase.ID,
> ```

I think all the logic for updating table descriptors could be moved further up the stack. That way this remains a simple function just for modifying the sequence descriptor, and it will allow for a clearer distinction between the CREATE SEQUENCE and ALTER SEQUENCE logic. (For instance the CREATE path doesn't need to worry about checking for a previous owner.) Also saves needing to plumb `runParams` all the way down here.

Let me know if there's some reason that is awkward. Btw, I think you asked me about this as you were writing it and I didn't understand the context well enough to advise - sorry about that. :)
___
*[pkg/sql/sequence.go, line 270 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-LpYlScE7dhlNkqWRBox:-LpYlScE7dhlNkqWRBoy:b9plxt2) ([raw file](https://github.com/cockroachdb/cockroach/blob/09c2833585e1c68d64dfb35c7deee384e02b01ca/pkg/sql/sequence.go#L270)):*
> ```Go
> 			p := *params.p
> 			ctx := params.ctx
> 			switch option.ColumnItemVal {
> ```

Style thing: Usually you don't see a switch statement here but rather `if option.ColumnItemVal == nil { ... } else { ... }`
___
*[pkg/sql/sequence.go, line 297 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-LpYlHCd0EL2pStBJ1OW:-LpYlHCd0EL2pStBJ1OX:b-bmtyni) ([raw file](https://github.com/cockroachdb/cockroach/blob/09c2833585e1c68d64dfb35c7deee384e02b01ca/pkg/sql/sequence.go#L297)):*
> ```Go
> 				}
> 				opts.SequenceOwner.HasOwner = false
> 			default:
> ```

Is there a case you need to handle here where the sequence was already owned by a different table/column?
___
*[pkg/sql/sequence.go, line 299 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-LpYkVXq42ZlmSteh3MH:-LpYkVXq42ZlmSteh3MI:bn4r3sr) ([raw file](https://github.com/cockroachdb/cockroach/blob/09c2833585e1c68d64dfb35c7deee384e02b01ca/pkg/sql/sequence.go#L299)):*
> ```Go
> 			default:
> 				var tableName tree.TableName
> 				if option.ColumnItemVal.TableName != nil {
> ```

What if it is nil?
___
*[pkg/sql/sequence.go, line 303 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-LpYkbPs67szmNCnCWQV:-LpYkbPs67szmNCnCWQW:b-dnp8ld) ([raw file](https://github.com/cockroachdb/cockroach/blob/09c2833585e1c68d64dfb35c7deee384e02b01ca/pkg/sql/sequence.go#L303)):*
> ```Go
> 				}
> 
> 				tableDesc, err := p.ResolveMutableTableDescriptor(ctx, &tableName, true, ResolveRequireTableDesc)
> ```

Style thing: replace `true` with `true /* required */` to help readability.
___
*[pkg/sql/sequence.go, line 307 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-LpYk2GM0a4XQ1Gl8XEa:-LpYk2GM0a4XQ1Gl8XEb:b-503pv1) ([raw file](https://github.com/cockroachdb/cockroach/blob/09c2833585e1c68d64dfb35c7deee384e02b01ca/pkg/sql/sequence.go#L307)):*
> ```Go
> 					return err
> 				}
> 				col, _, err := tableDesc.FindColumnByName(tree.Name(option.ColumnItemVal.ColumnName.String()))
> ```

Can you just replace `tree.Name(option.ColumnItemVal.ColumnName.String())` with `option.ColumnItemVal.ColumnName`?
___
*[pkg/sql/logictest/testdata/logic_test/sequences, line 953 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-LpYpchN5zF1Lbeg8QZa:-LpYpchOBV_q-qYPGDpZ:b-hiyp3h) ([raw file](https://github.com/cockroachdb/cockroach/blob/09c2833585e1c68d64dfb35c7deee384e02b01ca/pkg/sql/logictest/testdata/logic_test/sequences#L953)):*
> ```
>        JOIN pg_attribute AS attrib
>          ON ( attrib.attnum = dep.refobjsubid
>               AND attrib.attrelid = dep.refobjid )
> ```

Nit: All of the parens in this query are unnecessary
___
*[pkg/sql/logictest/testdata/logic_test/sequences, line 974 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-LpYpr1BBeAuDbFF93Q0:-LpYpr1BBeAuDbFF93Q1:b5c170i) ([raw file](https://github.com/cockroachdb/cockroach/blob/09c2833585e1c68d64dfb35c7deee384e02b01ca/pkg/sql/logictest/testdata/logic_test/sequences#L974)):*
> ```
>        JOIN pg_attribute AS attrib
>          ON ( attrib.attnum = dep.refobjsubid
>               AND attrib.attrelid = dep.refobjid )
> ```

What is the purpose of all the JOINs above? Seems like you could just check that the row is removed from `pg_class`. If you want to check that the rows are removed everywhere, you'll have to do something fancier because this query will return no rows as long as there is no row in `pg_class`.
___
*[pkg/sql/logictest/testdata/logic_test/sequences, line 995 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-LpYqNOq7bWaib1E9FZg:-LpYqNOq7bWaib1E9FZh:boxfgm9) ([raw file](https://github.com/cockroachdb/cockroach/blob/09c2833585e1c68d64dfb35c7deee384e02b01ca/pkg/sql/logictest/testdata/logic_test/sequences#L995)):*
> ```
> ALTER SEQUENCE owned_seq OWNED BY a.a;
> 
> statement error
> ```

Why is this an error?
___
*[pkg/sql/logictest/testdata/logic_test/sequences, line 1014 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-LpYqi4O2gNNQGlDSSOb:-LpYqi4O2gNNQGlDSSOc:b6jfe6y) ([raw file](https://github.com/cockroachdb/cockroach/blob/09c2833585e1c68d64dfb35c7deee384e02b01ca/pkg/sql/logictest/testdata/logic_test/sequences#L1014)):*
> ```
> 
> statement error cannot drop sequence id_seq because other objects depend on it
> ALTER TABLE ab DROP COLUMN a;
> ```

Where did the table `ab` come from?
___
*[pkg/sql/logictest/testdata/logic_test/sequences, line 1020 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-LpYqCe522uIARqQ-_PM:-LpYqCe522uIARqQ-_PN:b-kew4r3) ([raw file](https://github.com/cockroachdb/cockroach/blob/09c2833585e1c68d64dfb35c7deee384e02b01ca/pkg/sql/logictest/testdata/logic_test/sequences#L1020)):*
> ```
> 
> statement error relation "owned_seq" does not exist
> DROP SEQUENCE owned_seq;
> ```

Could use a test for changing the owner, if we support that. Also that you can't drop a table if another table depends on one of its owned sequences.


<!-- Sent from Reviewable.io -->

arulajmani reviewed (commented):
Thanks for the review! 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-:-LpZsqWn7Rz3h23c0usR:b-hlcaa9)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @solongordon)

___
*[pkg/ccl/importccl/read_import_mysql.go, line 269 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-LpYeirt2iVKKdqwaT2Z:-LpZKJLb2s0Kaxb_V-OC:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/09c2833585e1c68d64dfb35c7deee384e02b01ca/pkg/ccl/importccl/read_import_mysql.go#L269)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

By convention, `ctx` is usually the first argument.
</blockquote></details>

Done.
___
*[pkg/ccl/importccl/read_import_mysql.go, line 344 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-LpYfoyo3wKVBTglzSoo:-LpZLG16864b4xK1-GLx:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/09c2833585e1c68d64dfb35c7deee384e02b01ca/pkg/ccl/importccl/read_import_mysql.go#L344)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

I don't think this needs to be a pointer. `PlanHookState` is an interface, and the underlying implementation might be a pointer, but using a pointer to an interface is almost always undesirable in Go.
</blockquote></details>

Done.
___
*[pkg/ccl/importccl/read_import_pgdump.go, line 212 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-LpYgX0y0MKoMT1EZY5E:-LpZLQlv5UYlQqxH9aH0:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/09c2833585e1c68d64dfb35c7deee384e02b01ca/pkg/ccl/importccl/read_import_pgdump.go#L212)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

Ditto about making `ctx` the first argument.
</blockquote></details>

Done.
___
*[pkg/sql/sequence.go, line 191 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-LpYnPh586gXgtgOesNz:-LpZfFq73gu2CiMZf6jy:bvmtvqf) ([raw file](https://github.com/cockroachdb/cockroach/blob/09c2833585e1c68d64dfb35c7deee384e02b01ca/pkg/sql/sequence.go#L191)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

I think all the logic for updating table descriptors could be moved further up the stack. That way this remains a simple function just for modifying the sequence descriptor, and it will allow for a clearer distinction between the CREATE SEQUENCE and ALTER SEQUENCE logic. (For instance the CREATE path doesn't need to worry about checking for a previous owner.) Also saves needing to plumb `runParams` all the way down here.

Let me know if there's some reason that is awkward. Btw, I think you asked me about this as you were writing it and I didn't understand the context well enough to advise - sorry about that. :)
</blockquote></details>

As we talked about offline, I do not see an obvious way to get rid of passing down `runParams` until this layer because we need to save the Column/Table IDs in the sequenceDescriptor options, and the resolution from names to IDs requires access to `runParams`. 

I'm refactoring the removing previous owners / adding new owners to make this code more readable though as this function has gotten very clunky. 
___
*[pkg/sql/sequence.go, line 270 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-LpYlScE7dhlNkqWRBox:-LpZfxVuAXXXzVkwK98r:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/09c2833585e1c68d64dfb35c7deee384e02b01ca/pkg/sql/sequence.go#L270)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

Style thing: Usually you don't see a switch statement here but rather `if option.ColumnItemVal == nil { ... } else { ... }`
</blockquote></details>

Done.
___
*[pkg/sql/sequence.go, line 297 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-LpYlHCd0EL2pStBJ1OW:-LpZg0HCAzRBLw11La7w:b-qgeeft) ([raw file](https://github.com/cockroachdb/cockroach/blob/09c2833585e1c68d64dfb35c7deee384e02b01ca/pkg/sql/sequence.go#L297)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

Is there a case you need to handle here where the sequence was already owned by a different table/column?
</blockquote></details>

Good catch, I'm adding this as part of my refactor. 
___
*[pkg/sql/sequence.go, line 299 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-LpYkVXq42ZlmSteh3MH:-LpZLqIJ0H_3CFNC7DhZ:bi8ldg3) ([raw file](https://github.com/cockroachdb/cockroach/blob/09c2833585e1c68d64dfb35c7deee384e02b01ca/pkg/sql/sequence.go#L299)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

What if it is nil?
</blockquote></details>

I think the resolution below will fail and we will get an error. I copied this pattern from someplace else, I can't remember which file though. 

Technically, it should never be nil because we set this field during parsing, and if no table name is provided, we should get a syntax error.
___
*[pkg/sql/sequence.go, line 303 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-LpYkbPs67szmNCnCWQV:-LpZMc0eE3Qzq-trPPVQ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/09c2833585e1c68d64dfb35c7deee384e02b01ca/pkg/sql/sequence.go#L303)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

Style thing: replace `true` with `true /* required */` to help readability.
</blockquote></details>

Done.
___
*[pkg/sql/sequence.go, line 307 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-LpYk2GM0a4XQ1Gl8XEa:-LpZMwdRB17OhayH-szE:bs7wumx) ([raw file](https://github.com/cockroachdb/cockroach/blob/09c2833585e1c68d64dfb35c7deee384e02b01ca/pkg/sql/sequence.go#L307)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

Can you just replace `tree.Name(option.ColumnItemVal.ColumnName.String())` with `option.ColumnItemVal.ColumnName`?
</blockquote></details>

Done. 
___
*[pkg/sql/logictest/testdata/logic_test/sequences, line 974 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-LpYpr1BBeAuDbFF93Q0:-LpZhpn55-0XY3UzQWLP:b-lv4x6b) ([raw file](https://github.com/cockroachdb/cockroach/blob/09c2833585e1c68d64dfb35c7deee384e02b01ca/pkg/sql/logictest/testdata/logic_test/sequences#L974)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

What is the purpose of all the JOINs above? Seems like you could just check that the row is removed from `pg_class`. If you want to check that the rows are removed everywhere, you'll have to do something fancier because this query will return no rows as long as there is no row in `pg_class`.
</blockquote></details>

I think the row still exists in pg_class because the sequence has not been dropped yet. So `SELECT relname FROM pg_class WHERE relkind = 'S'` should return the sequence name. 
 
By checking this query does not return any rows, we are essentially ensuring that the `pg_depend` table no longer captures the sequence <-> table dependency once the owner has been removed. 

I could simplify this query a bit by removing the part that gets the column name out of the `pg_attribute` table as that is a redundant join.
___
*[pkg/sql/logictest/testdata/logic_test/sequences, line 995 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-LpYqNOq7bWaib1E9FZg:-LpZjRcO4xLSghYb75SO:bgvd35w) ([raw file](https://github.com/cockroachdb/cockroach/blob/09c2833585e1c68d64dfb35c7deee384e02b01ca/pkg/sql/logictest/testdata/logic_test/sequences#L995)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

Why is this an error?
</blockquote></details>

Because I was using the wrong command, the last two tests were never compiled and there's quite a few errors in them. I've fixed them all now and tested them with the correct command! My bad 
___
*[pkg/sql/logictest/testdata/logic_test/sequences, line 1014 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-LpYqi4O2gNNQGlDSSOb:-LpZjpe56cv8VUJaDkdO:b-8rg28o) ([raw file](https://github.com/cockroachdb/cockroach/blob/09c2833585e1c68d64dfb35c7deee384e02b01ca/pkg/sql/logictest/testdata/logic_test/sequences#L1014)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

Where did the table `ab` come from?
</blockquote></details>

Same reason as above. 
___
*[pkg/sql/logictest/testdata/logic_test/sequences, line 1020 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-LpYqCe522uIARqQ-_PM:-LpZjlMCCpEaFgUN39vW:b-pdtwqk) ([raw file](https://github.com/cockroachdb/cockroach/blob/09c2833585e1c68d64dfb35c7deee384e02b01ca/pkg/sql/logictest/testdata/logic_test/sequences#L1020)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

Could use a test for changing the owner, if we support that. Also that you can't drop a table if another table depends on one of its owned sequences.
</blockquote></details>

Added a test for ownership change and external table dependency as well. 

One thing to note on the external table dependency, the CASCADE option on the first table would error out -- we do not support DROP SEQUENCE CASCADE yet, and that is required to mimic postgres. 


<!-- Sent from Reviewable.io -->

solongordon reviewed (changes requested):
I think I got to everything now :crossed_fingers:.

> Reviewed 7 of 24 files at r3, 6 of 10 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-:-Lpe0A7t7aDWIXQ1ngZK:b-rikd1)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @solongordon)

___
*[pkg/ccl/importccl/read_import_mysql.go, line 395 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-LpcllP11pE7I6hIjJvR:-LpcllP21Et9xgWL5Md_:b-9w6uvh) ([raw file](https://github.com/cockroachdb/cockroach/blob/fad733f835560484346c029199d091b100eb964a/pkg/ccl/importccl/read_import_mysql.go#L395)):*
> ```Go
> 			params := p.RunParams(ctx)
> 			desc, err = sql.MakeSequenceTableDesc(
> 				seqName, opts, parentID, id, time, priv,nil /* settings */, &params,
> ```

Nit: Spacing looks a little funny here. If you haven't already, I'd recommend configuring a GoLand hook to run `crlfmt`: https://cockroachlabs.slack.com/archives/C4VM697PC/p1540908498001700
___
*[pkg/sql/drop_sequence.go, line 145 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-LpcqvwjAKnuLlZ7ZPdD:-LpcqvwjAKnuLlZ7ZPdE:b-q3k2o4) ([raw file](https://github.com/cockroachdb/cockroach/blob/fad733f835560484346c029199d091b100eb964a/pkg/sql/drop_sequence.go#L145)):*
> ```Go
> ) error {
> 	for _, sequenceID := range col.OwnsSequenceIds {
> 		seqDesc, err := p.Tables().getTableVersionByID(ctx, p.txn, sequenceID, tree.ObjectLookupFlags{})
> ```

Maybe better to use `p.LookupTableByID` here?
___
*[pkg/sql/drop_sequence.go, line 150 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-LpcrGwr6la97t0EAVoQ:-LpcrGwr6la97t0EAVoR:b-o01s52) ([raw file](https://github.com/cockroachdb/cockroach/blob/fad733f835560484346c029199d091b100eb964a/pkg/sql/drop_sequence.go#L150)):*
> ```Go
> 		}
> 		affectsNoColumns := len(seqDesc.DependedOnBy) == 0
> 		// It is okay if the sequence is depended on by columns that are being dropped om the same transaction
> ```

Nit: Try and wrap comments at 80 chars.
___
*[pkg/sql/pg_catalog.go, line 564 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-LpdelFl2t_q3372n62D:-LpdelFl2t_q3372n62E:b-3yxtcl) ([raw file](https://github.com/cockroachdb/cockroach/blob/fad733f835560484346c029199d091b100eb964a/pkg/sql/pg_catalog.go#L564)):*
> ```Go
> 					tree.DNull,                // relowner
> 					cockroachIndexEncodingOid, // relam
> 					relfilenode,               // relfilenode
> ```

I wonder if you need this change. Can you just use the `oid` column instead if you need the ID for joins? I'm wary about setting this field unless we're confident about the value. I tried doing `select oid, relfilenode from pg_class;` in Postgres and see many 0 values for this field.
___
*[pkg/sql/pg_catalog.go, line 1052 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-Lpdo4IK1CkG14pHYQ2g:-Lpdo4IK1CkG14pHYQ2h:b-5w555n) ([raw file](https://github.com/cockroachdb/cockroach/blob/fad733f835560484346c029199d091b100eb964a/pkg/sql/pg_catalog.go#L1052)):*
> ```Go
> 			pgClassTableOid := defaultOid(pgClassDesc.ID)
> 			if table.IsSequence() &&
> 				table.SequenceOpts.SequenceOwner != (sqlbase.TableDescriptor_SequenceOpts_SequenceOwner{}) {
> ```

For protobuf-generated structs like this one it's safer to use the `.Equal` method than `!=`. Probably doesn't make a difference in this case but it does if, for instance, some of your fields are slices.
___
*[pkg/sql/pg_catalog.go, line 1055 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-Lpdoxz72a_v0_ZEQrn8:-Lpdoxz72a_v0_ZEQrn9:b-fuz8tl) ([raw file](https://github.com/cockroachdb/cockroach/blob/fad733f835560484346c029199d091b100eb964a/pkg/sql/pg_catalog.go#L1055)):*
> ```Go
> 				tableID := table.SequenceOpts.SequenceOwner.OwnerTableID
> 				desc, err := tableLookup.getTableByID(tableID)
> 				if err != nil {
> ```

I wonder if it's better to log a warning here than to error out. It's bad if the table ID doesn't exist but maybe that shouldn't totally break this table. For instance, see the `"broken fk reference"` warning below.
___
*[pkg/sql/pg_catalog.go, line 1060 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-LpdtZ4P97NGWsy9gyCG:-LpdtZ4P97NGWsy9gyCH:b-4nijys) ([raw file](https://github.com/cockroachdb/cockroach/blob/fad733f835560484346c029199d091b100eb964a/pkg/sql/pg_catalog.go#L1060)):*
> ```Go
> 				columnIndexMap := desc.ColumnIdxMap()
> 				// Postgres expects columns to be 1 indexed, we store them as 0 indexed
> 				colPosition := columnIndexMap[table.SequenceOpts.SequenceOwner.OwnerColumnID] + 1
> ```

I think you just want the column ID here, not the column position. That's what we use for `pg_attribute.attnum`: https://github.com/cockroachdb/cockroach/blob/7df2a12d716db8a782922d8e611d05cb9f6d6b0a/pkg/sql/pg_catalog.go#L415

It's often the case that the ID is the position + 1, but not always, for instance when another column is dropped.

To see the discrepancy between your code and Postgres behavior, you can try this:
```
create table t (x int, y int);
create sequence s owned by t.y;
alter table t drop column x;
```
And then compare the resulting `pg_depend`s.
___
*[pkg/sql/pg_catalog.go, line 1066 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-LpdrSk9DBsKMNTguxoe:-LpdrSk9DBsKMNTguxof:b4gkx1n) ([raw file](https://github.com/cockroachdb/cockroach/blob/fad733f835560484346c029199d091b100eb964a/pkg/sql/pg_catalog.go#L1066)):*
> ```Go
> 					return err
> 				}
> 				refObjSubID := refObjSubIDBitArray.AsDInt(0)
> ```

I don't think this `DBitArray` dance is necessary. You should just be say `refObjSubID := tree.NewDInt(tree.DInt(colPosition))`.
___
*[pkg/sql/sequence.go, line 273 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-LpdyZ7f3QxROUX0tWxr:-LpdyZ7f3QxROUX0tWxs:bwjc3ho) ([raw file](https://github.com/cockroachdb/cockroach/blob/fad733f835560484346c029199d091b100eb964a/pkg/sql/sequence.go#L273)):*
> ```Go
> 			}
> 			ctx := params.ctx
> 			if err := removeSequenceOwnerIfExists(ctx, params, sequenceID, opts); err != nil {
> ```

Could we avoid calling this if the owner is already what we want it to be? That way we don't trigger unnecessary table schema changes.
___
*[pkg/sql/sequence.go, line 310 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-LpdxVHYA0cXDN1VRgNQ:-LpdxVHYA0cXDN1VRgNR:b-plmh4l) ([raw file](https://github.com/cockroachdb/cockroach/blob/fad733f835560484346c029199d091b100eb964a/pkg/sql/sequence.go#L310)):*
> ```Go
> func removeSequenceOwnerIfExists(
> 	ctx context.Context,
> 	params *runParams,
> ```

Perhaps just pass `p *planner` since that's all you're using?
___
*[pkg/sql/sequence.go, line 315 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-LpdxApo6PTDeeyibLJu:-LpdxApo6PTDeeyibLJv:b-1fyhlk) ([raw file](https://github.com/cockroachdb/cockroach/blob/fad733f835560484346c029199d091b100eb964a/pkg/sql/sequence.go#L315)):*
> ```Go
> ) error {
> 	p := params.p
> 	if opts.SequenceOwner != (sqlbase.TableDescriptor_SequenceOpts_SequenceOwner{}) {
> ```

Use `.Equal` here. Also, it's a bit cleaner to do this:
```
if opts.SequenceOwner.Equal(sqlbase.TableDescriptor_SequenceOpts_SequenceOwner{}) {
	return nil
}
```
That way your whole function isn't in an if block.
___
*[pkg/sql/sequence.go, line 316 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-Lpdy-oC7t_DvjZ6JUIj:-Lpdy-oC7t_DvjZ6JUIk:b-8tp606) ([raw file](https://github.com/cockroachdb/cockroach/blob/fad733f835560484346c029199d091b100eb964a/pkg/sql/sequence.go#L316)):*
> ```Go
> 	p := params.p
> 	if opts.SequenceOwner != (sqlbase.TableDescriptor_SequenceOpts_SequenceOwner{}) {
> 		tableDesc, err := p.Tables().getMutableTableVersionByID(ctx, opts.SequenceOwner.OwnerTableID, p.txn)
> ```

Again, probably better to use `p.LookupTableByID` here.
___
*[pkg/sql/logictest/testdata/logic_test/sequences, line 1068 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-LpdzahdABVNGmqkXUJY:-Lpdzahe9ltzMrTr5xh5:b-jvi2we) ([raw file](https://github.com/cockroachdb/cockroach/blob/fad733f835560484346c029199d091b100eb964a/pkg/sql/logictest/testdata/logic_test/sequences#L1068)):*
> ```
> CREATE TABLE b(b INT DEFAULT nextval('seq'));
> 
> statement error cannot drop sequence seq because other objects depend on it
> ```

Hm, this error message is not ideal. The user is trying to drop a table so it's confusing to complain about a sequence. Here's the Postgres error for this case:
```
ERROR:  cannot drop table a because other objects depend on it
DETAIL:  default for table b column b depends on sequence seq
HINT:  Use DROP ... CASCADE to drop the dependent objects too.
```
I don't think we necessarily need to replicate the DETAIL and HINT parts but it would be nice to match the ERROR.
___
*[pkg/sql/sqlbase/structured.proto, line 768 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-Lpe-IJ40FGgX1J__Kf9:-Lpe-IJ40FGgX1J__KfA:b-kon641) ([raw file](https://github.com/cockroachdb/cockroach/blob/fad733f835560484346c029199d091b100eb964a/pkg/sql/sqlbase/structured.proto#L768)):*
> ```protobuf
>                                           (gogoproto.casttype) = "ColumnID"];
>       // Sequence Owner's Table ID
>       optional uint32 owner_table_id = 3 [(gogoproto.nullable) = false,
> ```

Might as well update these IDs to 1 and 2.


<!-- Sent from Reviewable.io -->

arulajmani reviewed (commented):
TFTR! 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-:-LpfUScQ4JO-CsMFv9dt:bz5m8k4)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @solongordon)

___
*[pkg/sql/drop_sequence.go, line 145 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-LpcqvwjAKnuLlZ7ZPdD:-Lpf2hEP4XmOflf0YGOw:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fad733f835560484346c029199d091b100eb964a/pkg/sql/drop_sequence.go#L145)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

Maybe better to use `p.LookupTableByID` here?
</blockquote></details>

Done.
___
*[pkg/sql/pg_catalog.go, line 564 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-LpdelFl2t_q3372n62D:-LpfDfx771emBFIzUdtR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fad733f835560484346c029199d091b100eb964a/pkg/sql/pg_catalog.go#L564)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

I wonder if you need this change. Can you just use the `oid` column instead if you need the ID for joins? I'm wary about setting this field unless we're confident about the value. I tried doing `select oid, relfilenode from pg_class;` in Postgres and see many 0 values for this field.
</blockquote></details>

Done.
___
*[pkg/sql/pg_catalog.go, line 1052 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-Lpdo4IK1CkG14pHYQ2g:-LpfCAlg8BuqOi-wjhnc:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fad733f835560484346c029199d091b100eb964a/pkg/sql/pg_catalog.go#L1052)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

For protobuf-generated structs like this one it's safer to use the `.Equal` method than `!=`. Probably doesn't make a difference in this case but it does if, for instance, some of your fields are slices.
</blockquote></details>

Done.
___
*[pkg/sql/pg_catalog.go, line 1055 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-Lpdoxz72a_v0_ZEQrn8:-LpfBqG1D2z8vbJMai5v:b-4wee9m) ([raw file](https://github.com/cockroachdb/cockroach/blob/fad733f835560484346c029199d091b100eb964a/pkg/sql/pg_catalog.go#L1055)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

I wonder if it's better to log a warning here than to error out. It's bad if the table ID doesn't exist but maybe that shouldn't totally break this table. For instance, see the `"broken fk reference"` warning below.
</blockquote></details>

As we no longer use the columnIndexMap to find the column position, we do not need to load this table descriptor anymore. 
___
*[pkg/sql/pg_catalog.go, line 1060 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-LpdtZ4P97NGWsy9gyCG:-LpfB7JQ-lXRu1pPDbOO:bfo7bmp) ([raw file](https://github.com/cockroachdb/cockroach/blob/fad733f835560484346c029199d091b100eb964a/pkg/sql/pg_catalog.go#L1060)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

I think you just want the column ID here, not the column position. That's what we use for `pg_attribute.attnum`: https://github.com/cockroachdb/cockroach/blob/7df2a12d716db8a782922d8e611d05cb9f6d6b0a/pkg/sql/pg_catalog.go#L415

It's often the case that the ID is the position + 1, but not always, for instance when another column is dropped.

To see the discrepancy between your code and Postgres behavior, you can try this:
```
create table t (x int, y int);
create sequence s owned by t.y;
alter table t drop column x;
```
And then compare the resulting `pg_depend`s.
</blockquote></details>

I think I misunderstood the postgres docs, but this makes more sense + is way cleaner. Thanks! :+1: 
___
*[pkg/sql/pg_catalog.go, line 1066 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-LpdrSk9DBsKMNTguxoe:-LpfB53y-kjGLSAivC5E:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fad733f835560484346c029199d091b100eb964a/pkg/sql/pg_catalog.go#L1066)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

I don't think this `DBitArray` dance is necessary. You should just be say `refObjSubID := tree.NewDInt(tree.DInt(colPosition))`.
</blockquote></details>

Done.
___
*[pkg/sql/sequence.go, line 273 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-LpdyZ7f3QxROUX0tWxr:-LpfSEBw6pvQpnhvnk9E:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fad733f835560484346c029199d091b100eb964a/pkg/sql/sequence.go#L273)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

Could we avoid calling this if the owner is already what we want it to be? That way we don't trigger unnecessary table schema changes.
</blockquote></details>

Done.
___
*[pkg/sql/sequence.go, line 310 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-LpdxVHYA0cXDN1VRgNQ:-Lpf8YRD-CnV517bMauz:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fad733f835560484346c029199d091b100eb964a/pkg/sql/sequence.go#L310)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

Perhaps just pass `p *planner` since that's all you're using?
</blockquote></details>

Done.
___
*[pkg/sql/sequence.go, line 315 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-LpdxApo6PTDeeyibLJu:-LpfEGWZ-HxcfelIoMzf:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fad733f835560484346c029199d091b100eb964a/pkg/sql/sequence.go#L315)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

Use `.Equal` here. Also, it's a bit cleaner to do this:
```
if opts.SequenceOwner.Equal(sqlbase.TableDescriptor_SequenceOpts_SequenceOwner{}) {
	return nil
}
```
That way your whole function isn't in an if block.
</blockquote></details>

Done.
___
*[pkg/sql/sequence.go, line 316 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-Lpdy-oC7t_DvjZ6JUIj:-Lpf7FXHB4UhuLZAgUQv:b-glkdws) ([raw file](https://github.com/cockroachdb/cockroach/blob/fad733f835560484346c029199d091b100eb964a/pkg/sql/sequence.go#L316)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

Again, probably better to use `p.LookupTableByID` here.
</blockquote></details>

I don't think LookupTableByID will work here because we need a mutable descriptor, and LookupTableByID gives a TableEntry structure which contains an immutable descriptor. 
___
*[pkg/sql/logictest/testdata/logic_test/sequences, line 1068 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-LpdzahdABVNGmqkXUJY:-LpfGtak19qaD8UxkTSH:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fad733f835560484346c029199d091b100eb964a/pkg/sql/logictest/testdata/logic_test/sequences#L1068)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

Hm, this error message is not ideal. The user is trying to drop a table so it's confusing to complain about a sequence. Here's the Postgres error for this case:
```
ERROR:  cannot drop table a because other objects depend on it
DETAIL:  default for table b column b depends on sequence seq
HINT:  Use DROP ... CASCADE to drop the dependent objects too.
```
I don't think we necessarily need to replicate the DETAIL and HINT parts but it would be nice to match the ERROR.
</blockquote></details>

Done.
___
*[pkg/sql/sqlbase/structured.proto, line 768 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-Lpe-IJ40FGgX1J__Kf9:-Lpf1fJy1duV7J2JFlcN:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fad733f835560484346c029199d091b100eb964a/pkg/sql/sqlbase/structured.proto#L768)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

Might as well update these IDs to 1 and 2.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

solongordon reviewed (changes requested):
Some minor comments, but otherwise let's get this in!

Also, you can remove the "Release justification" from the commit message.

> Reviewed 3 of 6 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-:-LtRJdKr7orXSWCv1wkM:b-3p7lqb)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @solongordon)

___
*[pkg/ccl/importccl/import_stmt.go, line 473 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-LtUmNwd33lptkgt3dhJ:-LtUmNwd33lptkgt3dhK:bupdwa2) ([raw file](https://github.com/cockroachdb/cockroach/blob/c253bd4525bc37f899ab914ec9c409a271cc77f3/pkg/ccl/importccl/import_stmt.go#L473)):*
> ```Go
> 				case roachpb.IOFileFormat_PgDump:
> 					evalCtx := &p.ExtendedEvalContext().EvalContext
> 					tableDescs, err = readPostgresCreateTable(ctx, reader, evalCtx, p, p.ExecCfg().Settings, match, parentID, walltime, fks, int(format.PgDump.MaxRowSize))
> ```

Nit: Probably unnecessary to have the `settings` param now that you're passing `p`.
___
*[pkg/ccl/importccl/read_import_mysql.go, line 395 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-LtUmhJr5Oqi1q-7UWBV:-LtUmhJr5Oqi1q-7UWBW:bay86hh) ([raw file](https://github.com/cockroachdb/cockroach/blob/c253bd4525bc37f899ab914ec9c409a271cc77f3/pkg/ccl/importccl/read_import_mysql.go#L395)):*
> ```Go
> 			params := p.RunParams(ctx)
> 			desc, err = sql.MakeSequenceTableDesc(
> 				seqName, opts, parentID, id, time, priv,nil /* settings */, &params,
> ```

Weird whitespace
___
*[pkg/sql/alter_table.go, line 838 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-LtUnfxkC60f17pY82Gg:-LtUnfxkC60f17pY82Gh:blndo2m) ([raw file](https://github.com/cockroachdb/cockroach/blob/c253bd4525bc37f899ab914ec9c409a271cc77f3/pkg/sql/alter_table.go#L838)):*
> ```Go
> 			// Add references to the sequence descriptors this column is now using.
> 			changedSeqDescs, err := maybeAddSequenceDependencies(
> 				params.ctx, params.p, tableDesc, col, expr, nil,
> ```

Nit: Add a `/* backrefs */` comment
___
*[pkg/sql/create_sequence.go, line 138 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-LtUkWb44dpp9uwoD_TA:-LtUkWb44dpp9uwoD_TB:biv3a68) ([raw file](https://github.com/cockroachdb/cockroach/blob/c253bd4525bc37f899ab914ec9c409a271cc77f3/pkg/sql/create_sequence.go#L138)):*
> ```Go
> 	creationTime hlc.Timestamp,
> 	privileges *sqlbase.PrivilegeDescriptor,
> 	settings *cluster.Settings,
> ```

While you're here you might as well get rid of this param since it seems to be unused.
___
*[pkg/sql/create_table.go, line 1104 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-LtUovmT4u_-h4PqkmqC:-LtUovmT4u_-h4PqkmqD:b-n401mk) ([raw file](https://github.com/cockroachdb/cockroach/blob/c253bd4525bc37f899ab914ec9c409a271cc77f3/pkg/sql/create_table.go#L1104)):*
> ```Go
> 			if d.HasDefaultExpr() {
> 				// This resolution must be delayed until ColumnIDs have been populated.
> 				columnSequenceExprMap[len(desc.Columns)-1] = expr
> ```

Maybe change this name to `columnDefaultExprMap`, since `expr` doesn't necessarily use sequences.

It would also be a bit more idiomatic (at least as the Cockroach codebase goes) to make this a slice rather than a map. Then just append `expr` for every column, and the value will be nil if there's no default expression. This would also get rid of the awkward `len(...) - 1` business.
___
*[pkg/sql/drop_sequence.go, line 117 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-LtUqjGh769cEp7wkOCy:-LtUqjGh769cEp7wkOCz:b6x262n) ([raw file](https://github.com/cockroachdb/cockroach/blob/c253bd4525bc37f899ab914ec9c409a271cc77f3/pkg/sql/drop_sequence.go#L117)):*
> ```Go
> ) error {
> 	for _, col := range desc.Columns {
> 		if err := p.canRemoveOwnedSequencesImpl(ctx, desc, &col, behavior, false); err != nil {
> ```

Add `/* isColumnDrop */` param comment
___
*[pkg/sql/drop_sequence.go, line 130 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-LtUqlijBVQ2rPSmH7Nf:-LtUqlijBVQ2rPSmH7Ng:b6x262n) ([raw file](https://github.com/cockroachdb/cockroach/blob/c253bd4525bc37f899ab914ec9c409a271cc77f3/pkg/sql/drop_sequence.go#L130)):*
> ```Go
> 	behavior tree.DropBehavior,
> ) error {
> 	return p.canRemoveOwnedSequencesImpl(ctx, desc, col, behavior, true)
> ```

Add `/* isColumnDrop */` param comment
___
*[pkg/sql/drop_sequence.go, line 148 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-LtUrCQw1iCUtsD_5vxC:-LtUrCQw1iCUtsD_5vxD:b-rsk337) ([raw file](https://github.com/cockroachdb/cockroach/blob/c253bd4525bc37f899ab914ec9c409a271cc77f3/pkg/sql/drop_sequence.go#L148)):*
> ```Go
> 		affectsNoColumns := len(seqDesc.DependedOnBy) == 0
> 		// It is okay if the sequence is depended on by columns that are being
> 		// dropped om the same transaction
> ```

`om` typo
___
*[pkg/sql/pg_catalog.go, line 1055 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-LtUt6ehBmHxlh4_qOIr:-LtUt6ehBmHxlh4_qOIs:b-9xx3u1) ([raw file](https://github.com/cockroachdb/cockroach/blob/c253bd4525bc37f899ab914ec9c409a271cc77f3/pkg/sql/pg_catalog.go#L1055)):*
> ```Go
> 				refObjSubID := tree.NewDInt(tree.DInt(table.SequenceOpts.SequenceOwner.OwnerColumnID))
> 				objID := defaultOid(table.GetID())
> 				if err := addRow(
> ```

Nit: Just `return addRow(...`
___
*[pkg/sql/sequence.go, line 272 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-LtRK-r0-LQpefe2jnHW:-LtRK-r0-LQpefe2jnHX:b-xk8p0c) ([raw file](https://github.com/cockroachdb/cockroach/blob/c253bd4525bc37f899ab914ec9c409a271cc77f3/pkg/sql/sequence.go#L272)):*
> ```Go
> 					"Trying to add/remove Sequence Owner without access to context")
> 			}
> 			// If an owner is being removed
> ```

Nit: Move this comment into the if block and change to `The owner is being removed.`
___
*[pkg/sql/sequence.go, line 275 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-LtRJr8fES86b55ZSz1_:-LtRJr8g4BUsqDFwy1er:bcfh7dc) ([raw file](https://github.com/cockroachdb/cockroach/blob/c253bd4525bc37f899ab914ec9c409a271cc77f3/pkg/sql/sequence.go#L275)):*
> ```Go
> 			if option.ColumnItemVal == nil {
> 				if err := removeSequenceOwnerIfExists(params.ctx, params.p, sequenceID, opts); err != nil {
> 					return err
> ```

Nit: You can just `return removeSequenceOwnerIfExists(params.ctx, params.p, sequenceID, opts)`
___
*[pkg/sql/sequence.go, line 278 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-LtRK4cGCDRoD7jJ4IoM:-LtRK4cGCDRoD7jJ4IoN:b-szoe78) ([raw file](https://github.com/cockroachdb/cockroach/blob/c253bd4525bc37f899ab914ec9c409a271cc77f3/pkg/sql/sequence.go#L278)):*
> ```Go
> 				}
> 			} else {
> 				// If an owner is being added/modified
> ```

`// An owner is being added or modified.`
___
*[pkg/sql/sequence.go, line 288 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-LtRKNjM63_1OtMALSb-:-LtRKNjM63_1OtMALSb0:b-gkywj3) ([raw file](https://github.com/cockroachdb/cockroach/blob/c253bd4525bc37f899ab914ec9c409a271cc77f3/pkg/sql/sequence.go#L288)):*
> ```Go
> 				// We only want to trigger schema changes if the owner is not what we
> 				// want it to be.
> 				if opts.SequenceOwner.Equal(sqlbase.TableDescriptor_SequenceOpts_SequenceOwner{}) ||
> ```

Nit: I think this case is redundant. If `SequenceOwner` is empty then the following cases should be true right?
___
*[pkg/sql/sequence.go, line 359 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-LtUuKJe3XMjf2cFWIjL:-LtUuKJe3XMjf2cFWIjM:b-gghhxf) ([raw file](https://github.com/cockroachdb/cockroach/blob/c253bd4525bc37f899ab914ec9c409a271cc77f3/pkg/sql/sequence.go#L359)):*
> ```Go
> 	}
> 	// Reset the SequenceOwner to empty
> 	opts.SequenceOwner = sqlbase.TableDescriptor_SequenceOpts_SequenceOwner{}
> ```

Nit: You can call `opts.SequenceOwner.Reset()`.
___
*[pkg/sql/sequence.go, line 394 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-LtUvAjc4-HYqWRMirM2:-LtUvAjc4-HYqWRMirM3:b-7el8yc) ([raw file](https://github.com/cockroachdb/cockroach/blob/c253bd4525bc37f899ab914ec9c409a271cc77f3/pkg/sql/sequence.go#L394)):*
> ```Go
> 	opts.SequenceOwner.OwnerColumnID = col.ID
> 	opts.SequenceOwner.OwnerTableID = tableDesc.GetID()
> 	if err := p.writeSchemaChange(ctx, tableDesc, sqlbase.InvalidMutationID); err != nil {
> ```

Nit: Just return the result.
___
*[pkg/sql/logictest/testdata/logic_test/sequences, line 1027 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-LtUyS1X6aG3ggqoUHgD:-LtUyS1X6aG3ggqoUHgE:b-fa36jm) ([raw file](https://github.com/cockroachdb/cockroach/blob/c253bd4525bc37f899ab914ec9c409a271cc77f3/pkg/sql/logictest/testdata/logic_test/sequences#L1027)):*
> ```
> # cleanup
> statement ok
> SET sql_safe_updates = true;
> ```

Nit: I think it's fine to just turn off sql_safe_updates before your first test and leave it off.


<!-- Sent from Reviewable.io -->

arulajmani committed:
sql: Add support for OWNED BY option in CREATE/ALTER SEQUENCE Commands

This PR adds sequence ownership semantics to CRDB.

- By default, a sequence is not owned by any column.

- The OWNED BY option allows the user to establish an ownership
relationiship of a sequence  with a particular table's column.

- Ownership relationships can be removed by using "OWNED BY NONE"

Fixes #26382

Release note (sql change): CREATE/ALTER SEQUENCE now support the OWNED BY
syntax.

arulajmani reviewed (commented):
Thanks for the review. Should be good for another look :) 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-:-LtqFsTXBWeTot8wV8Ci:b-g7l2i3)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @solongordon)

___
*[pkg/ccl/importccl/read_import_mysql.go, line 395 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-LtUmhJr5Oqi1q-7UWBV:-Ltq3SVGDvM1vWuf83H6:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c253bd4525bc37f899ab914ec9c409a271cc77f3/pkg/ccl/importccl/read_import_mysql.go#L395)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

Weird whitespace
</blockquote></details>

Done.
___
*[pkg/sql/create_sequence.go, line 138 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-LtUkWb44dpp9uwoD_TA:-Ltq3f4m92wS5LWbpyCa:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c253bd4525bc37f899ab914ec9c409a271cc77f3/pkg/sql/create_sequence.go#L138)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

While you're here you might as well get rid of this param since it seems to be unused.
</blockquote></details>

Done.
___
*[pkg/sql/create_table.go, line 1104 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-LtUovmT4u_-h4PqkmqC:-LtqA7BSDEaAPXLfFa5j:b-549xr3) ([raw file](https://github.com/cockroachdb/cockroach/blob/c253bd4525bc37f899ab914ec9c409a271cc77f3/pkg/sql/create_table.go#L1104)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

Maybe change this name to `columnDefaultExprMap`, since `expr` doesn't necessarily use sequences.

It would also be a bit more idiomatic (at least as the Cockroach codebase goes) to make this a slice rather than a map. Then just append `expr` for every column, and the value will be nil if there's no default expression. This would also get rid of the awkward `len(...) - 1` business.
</blockquote></details>

Renamed to `columnDefaultExprs`, now that we have a slice instead of a map.  
___
*[pkg/sql/drop_sequence.go, line 117 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-LtUqjGh769cEp7wkOCy:-Ltq3y054J3jVemYr7xP:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c253bd4525bc37f899ab914ec9c409a271cc77f3/pkg/sql/drop_sequence.go#L117)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

Add `/* isColumnDrop */` param comment
</blockquote></details>

Done.
___
*[pkg/sql/drop_sequence.go, line 130 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-LtUqlijBVQ2rPSmH7Nf:-Ltq4D9J71N7WG2nDg6h:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c253bd4525bc37f899ab914ec9c409a271cc77f3/pkg/sql/drop_sequence.go#L130)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

Add `/* isColumnDrop */` param comment
</blockquote></details>

Done.
___
*[pkg/sql/drop_sequence.go, line 148 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-LtUrCQw1iCUtsD_5vxC:-Ltq4jmm2UwU77fDh7Pt:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c253bd4525bc37f899ab914ec9c409a271cc77f3/pkg/sql/drop_sequence.go#L148)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

`om` typo
</blockquote></details>

Done.
___
*[pkg/sql/sequence.go, line 275 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-LtRJr8fES86b55ZSz1_:-Ltq6vAbDGzu2TMkE1iL:bplnhlp) ([raw file](https://github.com/cockroachdb/cockroach/blob/c253bd4525bc37f899ab914ec9c409a271cc77f3/pkg/sql/sequence.go#L275)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

Nit: You can just `return removeSequenceOwnerIfExists(params.ctx, params.p, sequenceID, opts)`
</blockquote></details>

I think we can still have other options, that need to be handled if the error is not nil. for example, `CREATE SEQUENCE OWNED BY NONE START 2`
___
*[pkg/sql/sequence.go, line 278 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-LtRK4cGCDRoD7jJ4IoM:-Ltq7PAf3RtzxNhM_a6z:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c253bd4525bc37f899ab914ec9c409a271cc77f3/pkg/sql/sequence.go#L278)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

`// An owner is being added or modified.`
</blockquote></details>

Done.
___
*[pkg/sql/sequence.go, line 288 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-LtRKNjM63_1OtMALSb-:-Ltq7tnP0fkwNC1_wk-4:b2eszyi) ([raw file](https://github.com/cockroachdb/cockroach/blob/c253bd4525bc37f899ab914ec9c409a271cc77f3/pkg/sql/sequence.go#L288)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

Nit: I think this case is redundant. If `SequenceOwner` is empty then the following cases should be true right?
</blockquote></details>

I think I had this because I didn't want to issue a schema change if the owner, `table.a` of the sequence was being modified to be `table.a`. 


<!-- Sent from Reviewable.io -->

solongordon reviewed (approved):
> Reviewed 10 of 20 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/40992)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @solongordon)

___
*[pkg/sql/sequence.go, line 275 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-LtRJr8fES86b55ZSz1_:-Ltz-M-k-DekOo6Uz1X8:b9zvrap) ([raw file](https://github.com/cockroachdb/cockroach/blob/c253bd4525bc37f899ab914ec9c409a271cc77f3/pkg/sql/sequence.go#L275)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

I think we can still have other options, that need to be handled if the error is not nil. for example, `CREATE SEQUENCE OWNED BY NONE START 2`
</blockquote></details>

Oops, misread
___
*[pkg/sql/sequence.go, line 288 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-LtRKNjM63_1OtMALSb-:-Ltz-X8BEMcUt9I8fR0W:b-z33nb3) ([raw file](https://github.com/cockroachdb/cockroach/blob/c253bd4525bc37f899ab914ec9c409a271cc77f3/pkg/sql/sequence.go#L288)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

I think I had this because I didn't want to issue a schema change if the owner, `table.a` of the sequence was being modified to be `table.a`. 
</blockquote></details>

I was just talking about the `opts.SequenceOwner.Equal(sqlbase.TableDescriptor_SequenceOpts_SequenceOwner{})` check, but no big deal either way.


<!-- Sent from Reviewable.io -->

arulajmani reviewed (commented):
Turns out because we add the rowid column for tables without an explicit primary key, this was getting an index out of bounds error when I switched to the slice implementation from the map. Fixed that. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-:-LtzRpqPDFyhTKLbGQ2l:by4sykc)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @solongordon)



<!-- Sent from Reviewable.io -->

solongordon reviewed (approved):
Fix looks good!

> Reviewed 1 of 2 files at r7.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-:-LtzSkR-0vQsQ-DxyyzN:b-f2u2ew)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @solongordon)



<!-- Sent from Reviewable.io -->

arulajmani commented:
bors r+

knz commented:
Please fix CI before attempting merge.

(Or use `bors try`)

arulajmani reviewed (commented):
Turns out when this was failing on CI when I tried returning (_, error) in one of the functions defined in`sql.y`. Couldn't reproduce it locally, so not sure why that happens -- ended up having to remove the function, and duplicate a few lines.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/40992#-:-Lu8KS9BCaS2tLqQfFbo:b-xjdjlp)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @solongordon)



<!-- Sent from Reviewable.io -->

arulajmani commented:
bors r+
